### DISCUSSION TITLE
Providing additional error data for error translations

### DISCUSSION THREAD

**Discussion by tiii** [react-hook-form/react-hook-form#7977] [Q&A] [answered]
I'm currently showing sever validation errors via `setError`. Those errors come with additional data (e.g. `{ type: 'server', message:  'string-too-long', data: { maximum: 30 } }`) that is needed to generate a translated error message. If I would translate the error message before setting the error, it would not update when changing the locale.

`FieldError` does not allow additional data to be added (though you can add custom data to it, if you ignore the typing errors - obviously this is no viable option).

I will try to add the data to the context object and load the data from there when rendering the error, as I suppose this would be the way when using a custom resolver. Is there a simpler way to do this?

**Comment by tiii** [comment, marked as answer]
After searching for other terms, I figured that `react-hook-form` supports using react components as errors, though this is not reflected in the types. (#3654 and  #4289)

I now went that way type casting a react component as a string. This way the translation is self-contained including the necessary data to update when changing locale.

```tsx
setError(
  path,
  {
    type: reason,
    message: (
      <FormattedMessage
        id={`errors.${reason}`}
        values={data}
      />
    ) as unknown as string,
  },
  { shouldFocus: true }
);
```

**Reply by bryanprimus** [reply]
did you find another solution?
